include/rmm/mr/device/tracking_resource_adaptor.hpp

```diff
@@ -0,0 +1,261 @@
+    }
+#endif  // SPDLOG_ACTIVE_LEVEL <= SPDLOG_LEVEL_DEBUG
+  }
+
+ private:
+  /**
+   * @brief Allocates memory of size at least `bytes` using the upstream
+   * resource as long as it fits inside the allocation limit.
+   *
+   * The returned pointer has at least 256B alignment.
+   *
+   * @throws `rmm::bad_alloc` if the requested allocation could not be fulfilled
+   * by the upstream resource.
+   *
+   * @param bytes The size, in bytes, of the allocation
+   * @param stream Stream on which to perform the allocation
+   * @return void* Pointer to the newly allocated memory
+   */
+  void* do_allocate(std::size_t bytes, cudaStream_t stream) override
+  {
+    void* p = upstream_->allocate(bytes, stream);
+
+    // track it.
+    {
+      write_lock_t lock(mtx_);
+      allocations_.emplace(p, allocation_info{bytes, capture_stacks_});
+    }
+    allocated_bytes_ += bytes;
+
+    return p;
+  }
+
+  /**
+   * @brief Free allocation of size `bytes` pointed to by `p`
+   *
+   * @throws Nothing.
+   *
+   * @param p Pointer to be deallocated
+   * @param bytes Size of the allocation
+   * @param stream Stream on which to perform the deallocation
+   */
+  void do_deallocate(void* p, std::size_t bytes, cudaStream_t stream) override
+  {
+    upstream_->deallocate(p, bytes, stream);
+    {
+      write_lock_t lock(mtx_);
+      allocations_.erase(p);
+    }
+    allocated_bytes_ -= bytes;
+  }
+
+  /**
+   * @brief Compare the upstream resource to another.
+   *
+   * @throws Nothing.
+   *
+   * @param other The other resource to compare to
+   * @return true If the two resources are equivalent
+   * @return false If the two resources are not equal
+   */
+  bool do_is_equal(device_memory_resource const& other) const noexcept override
+  {
+    if (this == &other)
+      return true;
+    else {
+      tracking_resource_adaptor<Upstream> const* cast =
+        dynamic_cast<tracking_resource_adaptor<Upstream> const*>(&other);
+      return cast != nullptr ? upstream_->is_equal(*cast->get_upstream()) : upstream_->is_equal(other);
+    }
+  }
+
+  /**
+   * @brief Get free and available memory from upstream resource.
+   *
+   * @throws `rmm::cuda_error` if unable to retrieve memory info.
+   *
+   * @param stream Stream on which to get the mem info.
+   * @return std::pair contaiing free_size and total_size of memory
+   */
+  std::pair<std::size_t, std::size_t> do_get_mem_info(cudaStream_t stream) const override
+  {
+    return upstream_->get_mem_info(stream);
+  }
+
+  bool capture_stacks_;                           // whether or not to capture call stacks
+  std::map<void*, allocation_info> allocations_;  // map of active allocations
+  std::atomic<std::size_t> allocated_bytes_;      // number of bytes currently allocated
+  std::shared_timed_mutex mutable mtx_;           // mutex for thread safe access to allocations_
+  Upstream* upstream_;  ///< The upstream resource used for satisfying allocation requests
```

Comment:
```suggestion
  Upstream* upstream_;  // the upstream resource used for satisfying allocation requests
```

Reply:
I'm not familiar with this nomenclature. What does `/<` signify?

Reply:
Doxygen document the thing to the left.